types.go

```
@@ -41,6 +43,159 @@ func (o Strings) MarshalJSON() ([]byte, error) {
+
+const (
+	False   Flag = -1
+	Default Flag = 0
+	True    Flag = 1
+)
+
+// WithDefault resolves the value of the flag given the provided default value.
+func (f Flag) WithDefault(defaultValue bool) bool {
+	switch f {
+	case False:
+		return false
+	case Default:
+		return defaultValue
+	case True:
+		return true
+	default:
+		panic(fmt.Sprintf("invalid flag value %d", f))
+	}
+}
+
+func (f Flag) MarshalJSON() ([]byte, error) {
+	switch f {
+	case Default:
+		return json.Marshal(nil)
+	case True:
+		return json.Marshal(true)
+	case False:
+		return json.Marshal(false)
+	default:
+		return nil, fmt.Errorf("invalid flag value: %d", f)
+	}
+}
+
+func (f *Flag) UnmarshalJSON(input []byte) error {
+	switch string(input) {
+	case "null", "undefined":
+		*f = Default
+	case "false":
+		*f = False
+	case "true":
+		*f = True
+	default:
+		return fmt.Errorf("failed to unmarshal %q into a flag: must be null/undefined, true, or false", string(input))
+	}
+	return nil
+}
+
+func (f Flag) String() string {
+	switch f {
+	case Default:
+		return "default"
+	case True:
+		return "true"
+	case False:
+		return "false"
+	default:
+		return fmt.Sprintf("<invalid flag value %d>", f)
+	}
+}
+
+var _ json.Unmarshaler = (*Flag)(nil)
+var _ json.Marshaler = (*Flag)(nil)
+
+// Priority represents a value with a priority where 0 means "default" and -1
+// means "disabled".
+//
+// When encoded in json, Default is encoded as "null" and Disabled is encoded as
+// "false".
+type Priority int64
+
+const (
+	DefaultPriority Priority = 0
+	Disabled        Priority = -1
+)
+
+// WithDefault resolves the priority with the given default.
+//
+// If `defaultPriority` is Default/0, this function will return 0.
+func (p Priority) WithDefault(defaultPriority Priority) (priority int64, enabled bool) {
+	switch p {
+	case Disabled:
+		return 0, false
+	case DefaultPriority:
+		if defaultPriority < 0 {
+			return 0, false
+		}
+		return int64(defaultPriority), true
+	default:
+		return int64(p), true
```

Are we concerned at all about `p` being an invalid state with `p < -1`? If so this will return invalid data.

---

We could panic?

---

Seems reasonable. How do we tell users "hey you've entered an invalid config" that's not just silently failing? Is that also panicing, if so an error might be nicer,

---

We raise an error on decode.